apiserver/embeddedcommands.go

```diff
@@ -0,0 +1,231 @@
+)
+
+// ExecEmbeddedCommandFunc defines a function which runs a named Juju command
+// with the whitelisted sub commands.
+type ExecEmbeddedCommandFunc func(ctx *cmd.Context, store jujuclient.ClientStore, whitelist []string, cmd string, args []string) int
+
+// allowedEmbeddedCommands is a whitelist of Juju CLI commands which
+// are permissible to run embedded on a controller.
+var allowedEmbeddedCommands = []string{
+	"actions",
+	"add-machine",
+	"add-relation",
+	"add-space",
+	"add-storage",
+	"add-subnet",
+	"add-unit",
+	"add-user",
+	"agreements",
+	"attach",
+	"attach-resource",
+	"attach-storage",
+	"bind",
+	"cached-images",
+	"cancel-task",
+	"change-user-password",
+	"charm-resources",
+	"clouds",
+	"config",
+	"consume",
+	"controller-config",
+	"create-storage-pool",
+	"credentials",
+	"deploy",
+	"detach-storage",
+	"disable-user",
+	"enable-user",
+	"expose",
+	"find-offers",
+	"firewall-rules",
+	"get-constraints",
+	"get-model-constraints",
+	// TODO(wallyworld) - check if these should be allowed
+	//"grant",
+	//"grant-cloud",
+	"help",
+	"import-filesystem",
+	"machines",
+	"metrics",
+	"model-config",
+	"model-default",
+	"model-defaults",
+	"move-to-space",
+	"offer",
+	"offers",
+	"payloads",
+	"plans",
+	"register",
+	"relate",
+	"reload-spaces",
+	"remove-application",
+	"remove-cached-images",
+	"remove-consumed-application",
+	"remove-credential",
+	"remove-machine",
+	"remove-offer",
+	"remove-relation",
+	"remove-saas",
+	"remove-space",
+	"remove-storage",
+	"remove-storage-pool",
+	"remove-unit",
+	"remove-user",
+	"rename-space",
+	"resolved",
+	"resolve",
+	"resources",
+	"resume-relation",
+	"retry-provisioning",
+	"revoke",
+	"run",
+	"scale-application",
+	"set-credential",
+	"set-constraints",
+	"set-firewall-rule",
+	"set-meter-status",
+	"set-model-constraints",
+	"set-plan",
+	"set-series",
+	"set-wallet",
+	"show-action",
+	"show-application",
+	"show-cloud",
+	"show-controller",
+	"show-credential",
+	"show-credentials",
+	"show-machine",
+	"show-model",
+	"show-offer",
+	"show-status",
+	"show-status-log",
+	"show-storage",
+	"show-space",
+	"show-unit",
+	"show-user",
+	"show-wallet",
+	"sla",
+	"spaces",
+	"status",
+	"storage",
+	"storage-pools",
+	"subnets",
+	"suspend-relation",
+	"trust",
+	"unexpose",
+	"update-storage-pool",
+	"users",
+	"wallets",
+}
+
+// runCLICommands creates a CLI command instance with an in-memory copy of the controller,
+// model, and account details and runs the command against the host controller.
+func (srv *Server) runCLICommands(commands params.CLICommands) (params.StringResults, error) {
+	result := params.StringResults{}
+	if commands.AuthTag == "" {
+		return result, errors.NotSupportedf("CLI command for anonymous user")
+	}
+	userTag, err := names.ParseTag(commands.AuthTag)
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	if userTag.Kind() != names.UserTagKind {
+		return result, errors.NotSupportedf("CLI command for tag %q", userTag.Kind())
```

Comment:
tag shouldn't be exposed to users, it should be internal

Reply:
Issue here is that prams structs use tags.
model-tag is used also in the CLI args struct.
For this case, I could revert to using just UUID and username. It depends on whether we want to treat this http endpoint as an API call.